doc/tutorials/core/mat-mask-operations/mat_mask_operations.markdown

```diff
@@ -25,123 +28,169 @@ the zero-zero index) on the pixel you want to calculate and sum up the pixel val
+
+The function even has a fifth optional argument to specify the center of the kernel, a sixth
+for adding an optional value to the filtered pixels before storing them in K and a seventh one
+for determining what to do in the regions where the operation is undefined (borders).
+@end_toggle
+
+@add_toggle{Java}
+@snippet samples/java/tutorial_code/core/mat_mask_operations/MatMaskOperations.java kern
+
+Then call the [Imgproc.filter2D()] function specifying the input, the output image and the kernel to
+use:
+@snippet samples/java/tutorial_code/core/mat_mask_operations/MatMaskOperations.java filter2D
+The function even has a fifth optional argument to specify the center of the kernel, a sixth
+for adding an optional value to the filtered pixels before storing them in K and a seventh one
+for determining what to do in the regions where the operation is undefined (borders).
+@end_toggle
+
+@add_toggle{Python}
+@snippet samples/python/tutorial_code/core/mat_mask_operations/mat_mask_operations.py kern
+
+Then call the **cv2.filter2D()** function specifying the input, the output image and the kernell to
+use:
+@snippet samples/python/tutorial_code/core/mat_mask_operations/mat_mask_operations.py filter2D
+@end_toggle
+
+This function is shorter, less verbose and, because there are some optimizations, it is usually faster
+than the *hand-coded method*. For example in my test while the second one took only 13
+milliseconds the first took around 31 milliseconds. Quite some difference.
 
 For example:
 
 ![](images/resultMatMaskFilter2D.png)
 
+@add_toggle{C++}
 You can download this source code from [here
-](https://github.com/opencv/opencv/tree/master/samples/cpp/tutorial_code/core/mat_mask_operations/mat_mask_operations.cpp) or look in the
+](https://github.com/Itseez/opencv/tree/master/samples/cpp/tutorial_code/core/mat_mask_operations/mat_mask_operations.cpp) or look in the
 OpenCV source code libraries sample directory at
 `samples/cpp/tutorial_code/core/mat_mask_operations/mat_mask_operations.cpp`.
 
 Check out an instance of running the program on our [YouTube
 channel](http://www.youtube.com/watch?v=7PF1tAU9se4) .
+@youtube{https://www.youtube.com/watch?v=7PF1tAU9se4}
+@end_toggle
+
+@add_toggle{Java}
+You can look in the OpenCV source code libraries sample directory at
+`samples/java/tutorial_code/core/mat_mask_operations/MatMaskOperations.java`.
+@end_toggle
+
+@add_toggle{Python}
+You can look in the OpenCV source code libraries sample directory at
+`samples/python/tutorial_code/core/mat_mask_operations/mat_mask_operations.py`.
+@end_toggle
 
-\htmlonly
-<div align="center">
-<iframe width="560" height="349" src="https://www.youtube.com/embed/7PF1tAU9se4?hd=1" frameborder="0" allowfullscreen></iframe>
-</div>
-\endhtmlonly
+<!-- invisible references list -->
+[Imgproc.filter2D()]: http://docs.opencv.org/java/3.1.0/org/opencv/imgproc/Imgproc.html#filter2D-org.opencv.core.Mat-org.opencv.core.Mat-int-org.opencv.core.Mat-
```

Comment:
We need to implement additional documentation generation procedures for Java and Python and than add references to them. Currently we can not produce it, so it will be better to avoid such links.